### PR TITLE
Add checks

### PR DESCRIPTION
**Pull request by shikoku1**
Add checks for array types implemented in #33

### CODE REVIEW

File: astarte-device-sdk/AstarteDeviceSDK.cpp [hunk 1]
```diff
@@ -159,7 +160,15 @@ void AstarteDeviceSDK::createConsumer(const Hyperdrive::Interface &interface, co
...
-        mappingToType.insert(endpoint, typeStringToVariantType(typeString));
+
+        auto ty=typeStringToVariantType(typeString);
+        switch (ty.second) {
+            case endpointType::Scalar:
+            mappingToType.insert(endpoint, ty.first);
+
+            case endpointType::Array:
+            mappingToArrayType.insert(endpoint, ty.first);
+        }
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.cpp:171]: let's align code style to existing code style.

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-device-sdk/AstarteDeviceSDK.h [hunk 2]
```diff
@@ -97,7 +99,7 @@ protected Q_SLOTS:
     void createProducer(const Hyperdrive::Interface &interface, const QJsonObject &producerObject);
     void createConsumer(const Hyperdrive::Interface &interface, const QJsonObject &consumerObject);
 
-    QVariant::Type typeStringToVariantType(const QString &typeString) const;
+    std::pair<QVariant::Type, endpointType> typeStringToVariantType(const QString &typeString) const;
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.h:102]: We prefer QPair to std::pair

File: astarte-utils/AstarteGenericProducer.cpp [hunk 3]
```diff
@@ -129,14 +129,65 @@ bool AstarteGenericProducer::sendData(const QVariant &value, const QByteArray &t
 
 bool AstarteGenericProducer::sendData(const QList<QVariant> &value, const QByteArray &target, const QDateTime &timestamp, const QVariantHash &metadata)
 {
-    // TODO: checks
-    QHash<QByteArray, QByteArray> attributes;
-    attributes.insert("interfaceType", QByteArray::number(static_cast<int>(m_interfaceType)));
-    // TODO: handle reliability, retention and expiry
+    if (!target.startsWith('/') || target.endsWith('/') || target.contains("//")
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:132]: is all this code a duplicate of existing code? Can we reuse existing code / existing logic?

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-device-sdk/AstarteDeviceSDK.cpp [hunk 4]
```diff
@@ -159,7 +160,15 @@ void AstarteDeviceSDK::createConsumer(const Hyperdrive::Interface &interface, co
         mappingToTokens.insert(endpoint, tokens);
 
         QString typeString = mappingObj.value(QStringLiteral("type")).toString();
-        mappingToType.insert(endpoint, typeStringToVariantType(typeString));
+
+        auto ty=typeStringToVariantType(typeString);
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.cpp:164]: we don't use auto as part of our coding style, unless when strictly necessary.

File: astarte-device-sdk/AstarteDeviceSDK.h [hunk 5]
```diff
@@ -32,6 +32,8 @@ class AstarteGenericConsumer;
 class AstarteGenericProducer;
 class QJsonSchemaChecker;
 
+enum endpointType {Scalar, Array};
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.h:35]: `Scalar` and `Array` names are too generic, they might conflict with other stuff, lets use `AstarteScalarType` or stuff like that.

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-device-sdk/AstarteDeviceSDK.h [hunk 6]
```diff
@@ -32,6 +33,8 @@ class AstarteGenericConsumer;
 class AstarteGenericProducer;
 class QJsonSchemaChecker;
 
+enum endpointType {AstarteScalarType, AstarteArrayType};
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.h:36]: coding style

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-utils/AstarteGenericProducer.cpp [hunk 7]
```diff
@@ -41,12 +41,25 @@ QHash<QByteArray, QVariant::Type> AstarteGenericProducer::mappingToType() const
...
 }
 
-bool AstarteGenericProducer::sendData(const QVariant &value, const QByteArray &target,
-        const QDateTime &timestamp, const QVariantHash &metadata)
+QHash<QByteArray, QVariant::Type> AstarteGenericProducer::mappingToArrayType() const
+{
+    return m_mappingToArrayType;
+}
+
+bool validateTarget(const QByteArray &target)
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:49]: this function should be either static (with C meaning), or private and static (C++) meaning.

File: astarte-utils/AstarteGenericProducer.cpp [hunk 8]
```diff
@@ -41,12 +41,25 @@ QHash<QByteArray, QVariant::Type> AstarteGenericProducer::mappingToType() const
...
 {
     if (!target.startsWith('/') || target.endsWith('/') || target.contains("//")
             || target.contains(";") || target.contains('\n') || target.contains('\r')
             || target.contains('+') || target.contains('#')) {
+        return false;
+    }
+    return true;
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:56]: no need for `if (true) return true`; the expression result is already a boolean.

File: astarte-utils/AstarteGenericProducer.cpp (hunk 7 above)
Comment (bettio) [reply astarte-utils/AstarteGenericProducer.cpp:49]: this is a predicate, it should be called isValidTarget

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-utils/AstarteGenericProducer.cpp [hunk 9]
```diff
@@ -94,51 +103,61 @@ bool AstarteGenericProducer::sendData(const QVariant &value, const QByteArray &t
...
-                sendDataOnEndpoint(value.toLongLong(), target, attributes, timestamp, metadata);
-                return true;
-            case QVariant::String:
-                sendDataOnEndpoint(value.toString(), target, attributes, timestamp, metadata);
-                return true;
-            default:
-                qWarning() << "Can't find valid type for " << target;
+            } else {
+                // if there's something in the array
+                if (value.toList()[0].type() == m_mappingToArrayType.value(matchedMapping)) {
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:113]: coding style: we don't use `[index]`, we rather use `.at(index)`

File: astarte-utils/AstarteGenericProducer.cpp [hunk 10]
```diff
@@ -94,51 +103,61 @@ bool AstarteGenericProducer::sendData(const QVariant &value, const QByteArray &t
...
-            case QVariant::ByteArray:
-                sendDataOnEndpoint(value.toByteArray(), target, attributes, timestamp, metadata);
-                return true;
-            case QVariant::DateTime:
-                sendDataOnEndpoint(value.toDateTime(), target, attributes, timestamp, metadata);
-                return true;
-            case QVariant::Double:
-                sendDataOnEndpoint(value.toDouble(), target, attributes, timestamp, metadata);
+        if (value.type() == QVariant::List) {
+            if (value.toList().length() == 0){
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:107]: Is it necessary to have a special case for 0?

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-device-sdk/AstarteDeviceSDK.cpp [hunk 11]
```diff
@@ -302,7 +339,9 @@ bool AstarteDeviceSDK::sendData(const QByteArray &interface, const QVariantHash
     }
     // Verify mappings
     QHash< QByteArray, QVariant::Type > mappingToType = m_producers.value(interface)->mappingToType();
-    if (mappingToType.size() != value.size()) {
+    QHash< QByteArray, QVariant::Type > mappingToArrayType = m_producers.value(interface)->mappingToArrayType();
```
Comment (bettio) [inline astarte-device-sdk/AstarteDeviceSDK.cpp:342]: honestly I cannot understand what you are doing here. Can you explain it?

**Review by bettio — CHANGES_REQUESTED** [review]

File: astarte-device-sdk/AstarteDeviceSDK.cpp (hunk 11 above)
Comment (shikoku1) [reply astarte-device-sdk/AstarteDeviceSDK.cpp:342]: Array types are in a different QHash than the scalar types to not mix them

File: astarte-utils/AstarteGenericProducer.cpp [hunk 12]
```diff
@@ -94,51 +101,64 @@ bool AstarteGenericProducer::sendData(const QVariant &value, const QByteArray &t
...
+        if (value.type() == QVariant::List) {
+            QList<QVariant> valueList = value.toList();
+
+            // if the array is empty, we don't care
+            if (valueList.length() == 0){
+                sendDataOnEndpoint(QList<QVariant>({}), target, attributes, timestamp, metadata);
                 return true;
-            case QVariant::String:
-                sendDataOnEndpoint(value.toString(), target, attributes, timestamp, metadata);
+            }
```
Comment (bettio) [inline astarte-utils/AstarteGenericProducer.cpp:111]: let's use `} else {`

**Review by bettio — CHANGES_REQUESTED** [review]

**Review by bettio — APPROVED** [review]

**Pull request merged**